README.md
[![Build Status](https://api.travis-ci.org/vissense/vissense.png?branch=master)](https://travis-ci.org/vissense/vissense)
[![Coverage Status](https://coveralls.io/repos/vissense/vissense/badge.png)](https://coveralls.io/r/vissense/vissense)
[![Dependency Status](https://david-dm.org/vissense/vissense.svg)](https://david-dm.org/vissense/vissense)
[![devDependency Status](https://david-dm.org/vissense/vissense/dev-status.svg)](https://david-dm.org/vissense/vissense#info=devDependencies)
[![Inline docs](http://inch-ci.org/github/vissense/vissense.svg?branch=master)](http://inch-ci.org/github/vissense/vissense)
![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)

[![Sauce Test Status](https://saucelabs.com/browser-matrix/vissense.svg)](https://saucelabs.com/u/vissense)

![logo](https://raw.githubusercontent.com/vissense/vissense/master/vissense-logo.png)

VisSense.js
====

A utility library for observing visibility changes of DOM elements.
Immediately know when an element becomes hidden, partly visible or fully visible.

VisSense.js is **lightweight** (<4KB minified and gzipped), **tested** and **documented**. Best of all: **No dependencies**.

#### What it does
 * provides methods for detecting visibility of DOM elements
 * provides a convenience class with straightforward methods like `isHidden`, `isVisible`, `isFullyVisible`, `percentage`
 * provides a convenience class for detecting changes in visibility

#### What it does *not* do (by default)
 * detect if an element is overlapped by others
 * take elements opacity into account
 * take scrollbars into account - elements "hidden" behind scrollbars are considered visible


Demos and Examples
------------
Check out [this bin for a quick demo](https://jsbin.com/yoqopa/edit?js,console,output).
See more examples on the [demo page](https://vissense.github.io/vissense-demo/).

In this simple example a video will only be started if at least 75% of its area is in the users viewport:
```javascript
var video = $('#video'); 
var videoVisibility = VisSense(video[0]);

if(videoVisibility.percentage() > 0.75) { 
  video.play();
}
```

In the following example the video will be started if it's at least 75% visible and stopped as soon as it is not visible anymore:
```javascript
var video = $('#video'); 
var visibility = VisSense(video[0], { fullyvisible: 0.75 });

var visibility_monitor = visibility.monitor({
  fullyvisible: function() { 
    video.play();
  }, 
  hidden: function() { 
    video.stop(); 
  }
}).start();
```
See a slightly adapted version of this example live and [try it on jsbin.com](https://jsbin.com/maqaco/edit?js,output).

Documentation
------------
See [vissense.github.io/vissense](https://vissense.github.io/vissense/) or generate the documentation locally.

#### Generate documentation
Clone the repository and run `grunt docs`

Download
------------

### npm
Install with [npm](https://www.npmjs.com/package/vissense)
```
npm install vissense --save
```

### Bower
Install with bower
```
bower install vissense/vissense --save
```

### Github
[Download from Github](https://github.com/vissense/vissense/releases)


### cdnjs
[Reference from cdnjs.com](https://cdnjs.com/libraries/vissense)


Contribute
------------

- Issue Tracker: https://github.com/vissense/vissense/issues
- Source Code: https://github.com/vissense/vissense

### Clone Repository
`git clone https://github.com/vissense/vissense.git`

#### Install dependencies
```
npm install -g grunt-cli
npm install -g karma-cli
npm install -g bower
npm install
bower install
```

#### Build project

`grunt`

#### Run tests

`grunt test`

or

`grunt serve`

and it automatically opens `http://localhost:3000/SpecRunner.html` in your browser.


API
------------

### VisSense(element [, options])

Object constructor. Options:

- `hidden` (_default: 0_) - if percentage is equal or below this limit the element is considered hidden
- `fullyvisible` (_default: 1_) -  if percentage is equal or above this limit the element is considered fully visible

Note: you can omit `new` keyword when calling `VisSense(...)`

#### .percentage()

gets the current visible percentage (0..1)

#### .isHidden()

`true` if element is hidden

#### .isVisible()

`true` if element is visible

#### .isFullyVisible()

`true` if element is fully visible

#### .state()

returns an object representing the current state

```javascript
{ 
  "code": 0, 
  "state": "hidden", 
  "percentage": 0, 
  "visible": false, 
  "fullyvisible": false, 
  "hidden": true,
  "previous": {}
}
```

#### .monitor([config])

This is an alias for getting a VisSense.VisMon object observing the current element. 
See the options below for more details.

```javascript
var element = document.getElementById('video');

var visibility_monitor = VisSense(element).monitor({
  visible: function() { 
    console.log('visible');
  }, 
  hidden: function() { 
    console.log('hidden');
  }
}).start();
```

### VisSense.VisMon(visobj [, options])

A monitor object lets you observe an element over a period of time.
It emits certain events you can subscribe to, and it can be extended with custom logic.

Object constructor. Options:

- `strategy` a strategy (or array of strategies) for observing the element. VisSense comes with two predefined strategies. See below.
- `start` function to run when monitoring the element starts
- `stop` function to run when monitoring the element stops
- `update` function to run when elements update function is called
- `hidden` function to run when element becomes hidden
- `visible` function to run when element becomes visible
- `fullyvisible` function to run when element becomes fully visible
- `visibilitychange` function to run when the visibility of the element changes
- `percentagechange` function to run when the percentage of the element changes
- `async` a boolean flag indicating whether events are synchronous or asynchronous

```javascript
var visobj = VisSense(document.getElementById('video'));

var visibilityMonitor = VisSense.VisMon(visobj, { 
  strategy: [
    new VisSense.VisMon.Strategy.EventStrategy({ throttle: 42 })
  ],
  visibilitychange: function() { 
    console.log('visibilitychange');
  }, 
  visible: function() { 
    console.log('element became visible');
  }
}).start();
```

##### Strategies
Strategies are hooks which let you intercept the monitoring process.
e.g. updating the monitor, sending custom events, etc.

VisSense comes with two predefined strategies:
  - `PollingStrategy` a simple strategy which invokes `update()` periodically.
  - `EventStrategy` this strategy registers event handlers for `visibilitychange`, `scroll` and `resize` events and calls `update()` accordingly.

A monitor can use any number of strategies. The default monitor uses a combination of `EventStrategy` and `PollingStrategy`. Feel free to write your own strategy to cover your specific requirements (it's super easy).
You can also pass an empty array if you don't want to use any strategy.

#### .start() 
starts observing the element returns `this`

#### .stop() 
stops observing the element

#### .update() 
manually run the update procedure. this will fire all registered hooks accordingly e.g. when a percentage change happened.

#### .state() 
returns a state object

```javascript
{ 
  "code": 1, 
  "state": "visible", 
  "percentage": 0.55, 
  "visible": true, 
  "fullyvisible": false, 
  "hidden": false 
  "previous" : { 
    "code": 2, 
    "state": "fullyvisible", 
    "percentage": 1, 
    "visible": true, 
    "fullyvisible": true, 
    "hidden": false 
  }
} 
```

#### .on(event, hook) 
registers an event hook

```javascript
visibility_monitor.on('percentagechange', function() { ... });
```


##### Builder
There is a builder available if you want to build more complex monitor objects.

```javascript
var visobj = VisSense(document.getElementById('video'));

var visibilityMonitor = VisSense.VisMon.Builder(visobj)
  .strategy(new VisSense.VisMon.Strategy.ConfigurablePollingStrategy({
    hidden: 1000,
    visible: 2000,
    fullyvisible: 5000
  }))
  .strategy(new VisSense.VisMon.Strategy.EventStrategy({ throttle: 200 }))
  .strategy(new VisSense.VisMon.Strategy.UserActivityStrategy({ inactiveAfter: 60000 }))
  .strategy(new VisSense.VisMon.Strategy.PercentageTimeTestEventStrategy('50%/10s', { 
    percentageLimit: 0.5,
    timeLimit: 10000,
    interval: 100
  }))
  .strategy({
    start: function(monitor) {
      setTimeout(function() {
        monitor.publish('mySpecialEvent');
      }, 10000);
    }
  })
  .on('start', function (monitor) {
      console.log('[Visibility Monitor] Started!');
  })
  .on('stop', function (monitor) {
    console.log('[Visibility Monitor] Stopped!');
  })
  .on('visible', function (monitor) {
    console.log('[Visibility Monitor] Element became visible!');
  })
  .on('50%/10s', function (monitor) {
    console.log('[Visibility Monitor] Element was >50% visible for 10 seconds!');
  })
  .on('mySpecialEvent', function (monitor) {
    console.log('[Visibility Monitor] MySpecialEvent received!');
  })
  .build()
  .startAsync();
```
This example uses external strategies. See [UserActivityStrategy](https://github.com/vissense/vissense-user-activity), [PercentageTimeTestEventStrategy](https://github.com/vissense/vissense-percentage-time-test) and [ConfigurablePollingStrategy](https://github.com/vissense/vissense-configurable-polling-strategy) for more information.


License
-------

The project is licensed under the MIT license. See
[LICENSE](LICENSE) for details.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]

## [0.10.0] - 2016-02-09
### Changed
- monitor can only be updated if it has been started before

### Fixed
- possible multiple calls to monitor listeners after stop event
- publish 'stop' event if monitor is stopped in first update cycle

## [0.9.0] - 2016-01-12
### Added
- CommonJS compatibility
- added config parameter `precision` to VisSense config object

### Changed
- `Utils.percentage()` does not round return value

## [0.8.3] - 2015-09-07
### Fixed
- EventStrategy adds listeners to referenceWindow

## [0.8.2] - 2015-06-11
### Fixed
- check computed style attribute "visibility" only once

## [0.8.1] - 2015-05-17
### Changed
- removed the banner from the minified file (a comment in the first row of the file vissense.min.js)

## [0.8.0] - 2015-04-15
### Added
- added support for defining a VisSense objects reference window
- added function `Utils.createVisibilityApi`
- added method `VisSense.referenceWindow`

## [0.7.0] - 2015-04-13
### Added
- added method `VisMon.publish`
- added `VisMon.Builder`

## [0.6.1] - 2015-03-17
### Fixed
- prevent multiple executions of a throttled function

## [0.6.0] - 2015-03-16
### Added
- added function `Utils.throttle`

### Changed
- adapted EventStrategy to use `Utils.throttle` instead of `Utils.debounce`

### Deprecated
- deprecated EventStrategy option `debounce` - use `throttle` instead

## [0.5.0] - 2015-03-14
### Added
- added function `VisSense.noConflict`

### Removed
- removed deprecated method `VisMon.use`
- removed deprecated method `VisMon.onUpdate` - use `VisMon.on('update', ...)`
- removed deprecated method `VisMon.onHidden` - use `VisMon.on('hidden', ...)`
- removed deprecated method `VisMon.onVisible` - use `VisMon.on('visible', ...)`
- removed deprecated method `VisMon.onFullyVisible` - use `VisMon.on('fullyvisible', ...)`
- removed deprecated method `VisMon.onVisibilityChange` - use `VisMon.on('visibilitychange', ...)`
- removed deprecated method `VisMon.onPercentageChange` - use `VisMon.on('percentagechange', ...)`

### Fixed
- prevent emitting start/stop events if already started/stopped

## [0.4.0] - 2015-03-03
### Added
- added function `Utils.once`
- added method `VisSense.element`
- added method `Strategy.init`
- added `start` and `stop` events to `VisMon`
- added `Utils.VisibilityApi`
- react on `touchmove` events in `VisMon.EventStrategy`

### Changed
- changed order of `percentagechange` event parameters
- listeners on all events (including `*`) are only called once per even
- template methods `Strategy.start` and `Strategy.stop` do not throw an error by default

### Deprecated
- deprecated `Vismon.on<Event>` - use `Vismon.on('<Event>', ...)` instead
- deprecated `VisMon.use`


## [0.3.0] - 2015-02-08
### Added
- added `Utils.async`
- added `VisMon.startAsync`
- added constructor option `visibilityHooks` to `VisSense`

### Removed
- removed `VisMon.NoopStrategy`

[unreleased]: https://github.com/vissense/vissense/compare/0.10.0...HEAD
[0.10.0]: https://github.com/vissense/vissense/compare/0.9.0...0.10.0
[0.9.0]: https://github.com/vissense/vissense/compare/0.8.3...0.9.0
[0.8.3]: https://github.com/vissense/vissense/compare/0.8.2...0.8.3
[0.8.2]: https://github.com/vissense/vissense/compare/0.8.1...0.8.2
[0.8.1]: https://github.com/vissense/vissense/compare/0.8.0...0.8.1
[0.8.0]: https://github.com/vissense/vissense/compare/0.7.0...0.8.0
[0.7.0]: https://github.com/vissense/vissense/compare/0.6.1...0.7.0
[0.6.1]: https://github.com/vissense/vissense/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/vissense/vissense/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/vissense/vissense/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/vissense/vissense/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/vissense/vissense/compare/0.2.1...0.3.0


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  require('time-grunt')(grunt);

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    banner: '/*! { ' +
    '"name": "<%= pkg.name %>", ' +
    '"version": "<%= pkg.version %>", ' +
    '<%= pkg.homepage ? "\\"homepage\\": \\"" + pkg.homepage + "\\"," : "" %>' +
    '"copyright": "(c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>" ' +
    '} */\n',

    dirs: {
      tmp: './tmp',
      dist: './dist',
      src: './lib',
      docs: './dist/docs',
      coverage: '<%= dirs.dist %>/coverage'
    },
    clean: {
      tmp: {
        src: ['<%= dirs.tmp %>']
      },
      docs: {
        src: ['<%= dirs.docs %>']
      },
      dist: {
        src: ['<%= dirs.dist %>']
      }
    },
    concat: {
      tmp: {
        src: [
          '<%= dirs.src %>/<%= pkg.name %>.js'
        ],
        dest: '<%= dirs.tmp %>/<%= pkg.name %>.js'
      }
    },
    umd: {
      all: {
        options: {
          src: '<%= concat.tmp.dest %>',
          dest: '<%= concat.tmp.dest %>',
          template: './templates/umd-vissense-no-conflict.hbs',
          objectToExport: 'VisSense',
          indent: 2
        }
      }
    },
    uglify: {
      src: {
        options: {
          banner: '<%= banner %>',
          compress: {
            drop_console: true
          },
          sourceMap: false,
          preserveComments: false,
          beautify: true,
          mangle: false
        },
        src: '<%= concat.tmp.dest %>',
        dest: '<%= dirs.dist %>/<%= pkg.name %>.js'
      },
      dist: {
        options: {
          banner: '',
          report: 'gzip',
          compress: {
            drop_console: true
          },
          sourceMap: false
        },
        src: '<%= concat.tmp.dest %>',
        dest: '<%= dirs.dist %>/<%= pkg.name %>.min.js'
      }
    },
    jshint: {
      options: {
        jshintrc: true
      },
      gruntfile: {
        src: 'Gruntfile.js'
      },
      karma: {
        src: ['karma.conf.js', 'karma-saucelabs.conf.js']
      },
      src_test: {
        src: ['tmp/**/*.js', 'lib/**/*.js', 'spec/**/*.js']
      }
    },
    connect: {
      server: {
        options: {
          keepalive: true,
          hostname: 'localhost',
          port: 3000,
          base: './',
          livereload: true,
          open: 'http://localhost:3000/SpecRunner.html'
        }
      },
      docs: {
        options: {
          open: true,
          keepalive: true,
          hostname: 'localhost',
          port: 3000,
          base: '<%= dirs.docs %>'
        }
      }
    },
    watch: {
      gruntfile: {
        files: '<%= jshint.gruntfile.src %>',
        tasks: ['jshint:gruntfile']
      },
      src_test: {
        files: '<%= jshint.src_test.src %>',
        tasks: ['jshint', 'default']
      },
      src: {
        files: ['<%= dirs.src %>/vissense.js'],
        tasks: ['jshint', 'dist']
      },
      livereload: {
        options: {livereload: true},
        files: ['<%= dirs.dist %>/vissense.js']
      }
    },
    bump: {
      options: {
        files: ['package.json', 'bower.json'],
        updateConfigs: [],
        commit: true,
        commitMessage: 'release v%VERSION%',
        commitFiles: ['-a'],
        createTag: true,
        tagName: '%VERSION%',
        tagMessage: 'version %VERSION%',
        push: false,
        pushTo: 'upstream',
        gitDescribeOptions: '--tags --always --abbrev=1 --dirty=-d'
      }
    },

    jasmine: {
      coverage: {
        src: [
          '<%= dirs.src %>/<%= pkg.name %>.js'
        ],
        options: {
          display: 'full',
          summary: true,
          specs: ['spec/*Spec.js'],
          helpers: [
            'spec/*Helper.js'
          ],
          vendor: [
            'bower_components/jquery/dist/jquery.min.js',
            'bower_components/lodash/dist/lodash.min.js',
            'bower_components/jasmine-jquery/lib/jasmine-jquery.js',
            'spec/phantomjsPolyfills.js'
          ],
          template: require('grunt-template-jasmine-istanbul'),
          templateOptions: {
            coverage: '<%= dirs.coverage %>/coverage.json',
            thresholds: {
              lines: 95,
              statements: 90,
              branches: 90,
              functions: 100
            },
            report: [{
              type: 'lcov',
              options: {
                dir: '<%= dirs.coverage %>/lcov'
              }
            }, {
              type: 'text-summary'
            }
            ]
          }
        }
      }
    },
    karma: {
      unit: {
        configFile: 'karma.conf.js'
      }
    },
    micro: {
      dist: {
        src: '<%= uglify.dist.dest %>',
        options: {
          limit: 4096, // default is 5KB
          gzip: true
        }
      }
    },
    complexity: {
      generic: {
        src: ['lib/vissense.js'],
        exclude: [],
        options: {
          breakOnErrors: false,
          errorsOnly: false,
          cyclomatic: [7], // [3, 7, 12]
          halstead: [25], // [8, 13, 20]
          maintainability: 100,
          hideComplexFunctions: false,
          broadcast: false
        }
      }
    },
    coveralls: {
      options: {
        force: true
      },
      target: {
        src: '<%= dirs.coverage %>/lcov/lcov.info'
      }
    },
    jsdoc: {
      docstrap: {
        src: ['lib/**/*.js', 'README.md'],
        options: {
          destination: '<%= dirs.docs %>',
          template: 'node_modules/ink-docstrap/template',
          configure: 'jsdoc-docstrap.conf.json'
        }
      }
    },
    notify: {
      js: {
        options: {
          title: 'Javascript - <%= pkg.title %>',
          message: 'Minified and validated with success!'
        }
      },
      test: {
        options: {
          title: 'Javascript - <%= pkg.title %>',
          message: 'Tests successfully finished!'
        }
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-connect');
  grunt.loadNpmTasks('grunt-contrib-jasmine');
  grunt.loadNpmTasks('grunt-contrib-clean');

  grunt.loadNpmTasks('grunt-karma');
  grunt.loadNpmTasks('grunt-micro');
  grunt.loadNpmTasks('grunt-coveralls');
  grunt.loadNpmTasks('grunt-bump');
  grunt.loadNpmTasks('grunt-notify');
  grunt.loadNpmTasks('grunt-umd');
  grunt.loadNpmTasks('grunt-jsdoc');
  grunt.loadNpmTasks('grunt-complexity');

  grunt.registerTask('dist', [
    'clean:tmp', 'clean:dist',
    'concat:tmp',
    'umd',
    'jshint',
    'complexity',
    'uglify',
    'micro',
    'clean:tmp']);

  grunt.registerTask('default', ['dist', 'test', 'notify:js']);

  grunt.registerTask('fast', ['dist', 'test-fast', 'notify:js']);

  grunt.registerTask('serve', ['dist', 'connect:server']);
  grunt.registerTask('server', ['serve']);

  grunt.registerTask('test', ['jasmine', 'karma', 'notify:test']);
  grunt.registerTask('test-fast', ['jasmine', 'notify:test']);

  grunt.registerTask('docs', ['clean:docs', 'jsdoc:docstrap', 'connect:docs']);
  grunt.registerTask('docs-only', ['clean:docs', 'jsdoc:docstrap']);

  grunt.registerTask('coverage', ['coveralls']);

};

